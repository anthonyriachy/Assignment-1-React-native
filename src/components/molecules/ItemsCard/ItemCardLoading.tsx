import { View } from 'react-native';
import ShimmerPlaceholder from 'react-native-shimmer-placeholder';
import { useTheme } from '../../../hooks/UseTheme';
import { ThemeColors } from '../../../constants/theme';
import LinearGradient from 'react-native-linear-gradient';
import { styles } from './ItemsCard.style';
export const ItemsCardShimmer = () => {
	const { colors } = useTheme() as { colors: ThemeColors };
	return (
		<View style={styles.parentContainer}>

				<ShimmerPlaceholder
					style={{ width: '100%', height: 129, borderTopLeftRadius: 15, borderTopRightRadius: 15 }}
					LinearGradient={LinearGradient}
				/>

					<ShimmerPlaceholder
						style={{ width: '40%', height: 20, marginBottom: 8 }}
					LinearGradient={LinearGradient}
					/>
					<ShimmerPlaceholder
						style={{ width: '80%', height: 20, marginBottom: 8 }}
						LinearGradient={LinearGradient}
					/>
					<ShimmerPlaceholder
						style={{ width: '60%', height: 20 }}
					LinearGradient={LinearGradient}
					/>


		</View>
	);
};
