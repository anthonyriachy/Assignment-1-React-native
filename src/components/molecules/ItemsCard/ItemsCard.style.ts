import { StyleSheet } from 'react-native';
import { window } from '../../../constants/sizes'
export const styles = StyleSheet.create({
	parentContainer: {
		height:255,
		width:window.width * 0.45+4,

	},
	container: {
		width:window.width * 0.45,
		height:250,
		borderRadius:15,
		elevation:3,
		shadowColor: '#000',
		shadowOffset: {
			width: 0,
			height: 1,
		},
		shadowOpacity: 0.25,
		shadowRadius: 2.84,
	},
});
