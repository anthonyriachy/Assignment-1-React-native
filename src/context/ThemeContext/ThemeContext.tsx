import React, { createContext, useEffect, useState } from 'react';
import { useColorScheme } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { ThemeType, getThemeColors } from '../../constants/theme';
import { ThemeContextType } from './ThemeContext.type';

export const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const systemColorScheme = useColorScheme();
  const [theme, setTheme] = useState<ThemeType>(systemColorScheme || 'light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as ThemeType);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    AsyncStorage.setItem('theme', newTheme);
  };

  const value = {
    theme,
    colors: getThemeColors(theme),
    toggleTheme,
    setTheme: (newTheme: ThemeType) => {
      setTheme(newTheme);
      AsyncStorage.setItem('theme', newTheme);
    },
  };

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
};

