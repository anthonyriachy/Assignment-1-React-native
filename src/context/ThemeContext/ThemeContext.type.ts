import { getThemeColors, ThemeType } from '../../constants/theme';

export type ThemeContextType = {
    theme: ThemeType;
    colors: ReturnType<typeof getThemeColors>;
    toggleTheme: () => void;
    setTheme: (theme: ThemeType) => void;
  }