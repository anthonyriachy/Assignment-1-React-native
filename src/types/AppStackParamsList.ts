import { AppStackRoutes } from '../constants/AppStackRoutes';

export type AppStackParamsList = {
    [AppStackRoutes.BottomTabs]: undefined;
    [AppStackRoutes.Details]: undefined;
    [AppStackRoutes.Products]: { title: string };
    [AppStackRoutes.Profile]: undefined;
    [AppStackRoutes.SellModal]: undefined;
    SellModal: undefined;
};