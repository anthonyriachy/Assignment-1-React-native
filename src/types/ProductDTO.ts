 
export type ProductDTO={
    location:LocationDTO;
	_id:string;
	title:string;
	description:string;
	price:number;
	images:ImageDTO[];
	user:{
		_id:string;
		email:string;
	}
	createdAt:string;
	updatedAt:string;
	score:number;
	
}

export type LocationDTO={		
	name:string;
	longitude:number;
	latitude:number;
 }

export type ImageDTO={
	url:string;
	_id:string;
}