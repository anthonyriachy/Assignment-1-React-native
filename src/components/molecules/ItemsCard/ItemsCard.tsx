import React from 'react';
import { ItemsCardImage } from '../../atoms/ItemsCardImage';
import { Pressable, View } from 'react-native';
import { styles } from './ItemsCard.style.ts';
import { ItemsCardInfo } from '../../atoms/ItemsCardInfo/ItemsCardInfo.tsx';
import { ItemsCardProps } from './ItemsCard.type';
import { useNavigation } from '@react-navigation/native';
import { CommonActions } from '@react-navigation/native';
import { getImageUrl } from '../../../lib/imageUtils';
import { useCallback } from 'react';

export function ItemsCard({item, smaller}: ItemsCardProps) {
	const navigation = useNavigation<any>();

	const handleClick = useCallback(() => {
		navigation.dispatch(
			CommonActions.navigate({
				name: 'Details',
				params: { itemId: item._id },
			})
		);
	}, [item._id, navigation]);

	return (
		<Pressable onPress={handleClick} style={styles.parentContainer} testID="items-card-pressable">
			<View style={!smaller ? styles.container : styles.smallerContainer} testID="items-card-container">
				<ItemsCardImage image={getImageUrl(item.images[0].url)} testID="items-card-image" item={item} />
				<ItemsCardInfo 
					title={item.title} 
					price={item.price} 
					date={item.createdAt} 
					location={item.location} 
				/>
			</View>
		</Pressable>
	);
}

