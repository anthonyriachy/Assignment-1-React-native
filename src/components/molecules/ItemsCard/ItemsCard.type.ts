import { ProductDTO } from "../../../types/ProductDTO";

export type ItemsCardProps = {
    item: ProductDTO;
    smaller?: boolean;
};