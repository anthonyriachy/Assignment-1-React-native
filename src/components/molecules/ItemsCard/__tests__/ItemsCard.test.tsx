// src/components/molecules/ItemsCard/__tests__/ItemsCard.test.tsx

import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { ItemsCard } from '../ItemsCard';
import { ThemeProvider } from '../../../../context/ThemeContext/ThemeContext';
import { ProductDTO } from '../../../../types/ProductDTO';

// ─── M O C K   N A V I G A T I O N ───────────────────────────────────────────────
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn().mockReturnValue({ dispatch: jest.fn() }),
  CommonActions: { navigate: jest.fn() },
}));

// ─── M O C K   U S E T H E M E ────────────────────────────────────────────────────
jest.mock('../../../../hooks/UseTheme', () => ({
  useTheme: () => ({
    colors: { background: '#fff', text: '#000', primary: '#007AFF', border: '#ccc' },
    theme: 'light',
  }),
}));

// ─── M O C K   F A S T I M A G E ───────────────────────────────────────────────────
jest.mock('react-native-fast-image', () => {
  const FastImage = () => null;
  FastImage.priority = { low: 0, normal: 1, high: 2 };
  FastImage.resizeMode = { contain: 'contain', cover: 'cover', stretch: 'stretch', center: 'center' };
  return FastImage;
});

// ─── M O C K   S H I M M E R   &   L I N E A R G R A D I E N T ────────────────────
jest.mock('react-native-shimmer-placeholder', () => 'ShimmerPlaceholder');
jest.mock('react-native-linear-gradient', () => 'LinearGradient');

// ─── M O C K   S V G   I C O N S ───────────────────────────────────────────────────
jest.mock('../../../../assets/icons/Heart.svg', () => 'HeartIcon');
jest.mock('../../../../assets/icons/DateIcon.svg', () => 'DateIcon');
jest.mock('../../../../assets/icons/LocationIcon.svg', () => 'LocationIcon');

// ─── M O C K   C H I L D   C O M P O N E N T S ────────────────────────────────────
// Note the three ".." to get from __tests__ → atoms
jest.mock('../../../atoms/ItemsCardImage', () => {
  const React = require('react');
  const { View, Text } = require('react-native');
  return {
    ItemsCardImage: ({ testID }: { testID?: string }) => (
      <View testID={testID}>
        <Text>MockImage</Text>
      </View>
    ),
  };
});

jest.mock('../../../atoms/ItemsCardInfo/ItemsCardInfo', () => {
  const React = require('react');
  const { View, Text } = require('react-native');
  return {
    ItemsCardInfo: ({ title, location }: any) => (
      <View testID="mock-info">
        <Text>{title}</Text>
        <Text>{location.name}</Text>
      </View>
    ),
  };
});

describe('ItemsCard', () => {
  const mockItem: ProductDTO = {
    _id: '123',
    title: 'Test Item',
    description: 'Test Description',
    price: 99.99,
    createdAt: '2024-03-20',
    updatedAt: '2024-03-20',
    location: { name: 'Test Location', longitude: 0, latitude: 0 },
    images: [{ url: 'test.jpg', _id: 'img1' }],
    user: { _id: 'user1', email: 'test@test.com' },
    score: 0,
  };

  it('renders correctly with default props', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <ItemsCard item={mockItem} />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly in smaller variant', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <ItemsCard item={mockItem} smaller={true} />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders without crashing and shows title & location', () => {
    const { getByTestId, getByText } = render(
      <ThemeProvider>
        <ItemsCard item={mockItem} />
      </ThemeProvider>
    );

    expect(getByTestId('items-card-pressable')).toBeTruthy();
    expect(getByTestId('items-card-image')).toBeTruthy();
    expect(getByText(/Test Item/)).toBeTruthy();
    expect(getByText(/Test Location/)).toBeTruthy();
  });

  it('dispatches navigation when pressed', () => {
    const mockDispatch = jest.fn();
    (require('@react-navigation/native').useNavigation as jest.Mock).mockReturnValue({
      dispatch: mockDispatch,
    });

    const { getByTestId } = render(
      <ThemeProvider>
        <ItemsCard item={mockItem} />
      </ThemeProvider>
    );

    fireEvent.press(getByTestId('items-card-pressable'));

    const { CommonActions } = require('@react-navigation/native');
    expect(mockDispatch).toHaveBeenCalledWith(
      CommonActions.navigate({ name: 'Details', params: { itemId: '123' } })
    );
  });
});
