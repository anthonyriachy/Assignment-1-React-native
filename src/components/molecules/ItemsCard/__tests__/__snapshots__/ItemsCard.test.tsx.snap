// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsCard renders correctly in smaller variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "width": 341.5,
    }
  }
  testID="items-card-pressable"
>
  <View
    style={
      {
        "borderRadius": 15,
        "elevation": 3,
        "height": 250,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 2.84,
        "width": 330,
      }
    }
    testID="items-card-container"
  >
    <View
      testID="items-card-image"
    >
      <Text>
        MockImage
      </Text>
    </View>
    <View
      testID="mock-info"
    >
      <Text>
        Test Item
      </Text>
      <Text>
        Test Location
      </Text>
    </View>
  </View>
</View>
`;

exports[`ItemsCard renders correctly with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "width": 341.5,
    }
  }
  testID="items-card-pressable"
>
  <View
    style={
      {
        "borderRadius": 15,
        "elevation": 3,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 2.84,
        "width": 341.5,
      }
    }
    testID="items-card-container"
  >
    <View
      testID="items-card-image"
    >
      <Text>
        MockImage
      </Text>
    </View>
    <View
      testID="mock-info"
    >
      <Text>
        Test Item
      </Text>
      <Text>
        Test Location
      </Text>
    </View>
  </View>
</View>
`;
